File: src/consume-keyword.js
/**
 * @fileoverview
 * Consumes a JSON keyword from JSONC input.
 */

/**
 * @typedef {import('./tokenize-jsonc.js').Token} Token
 * @typedef {import('./tokenize-jsonc.js').TokenizeError} TokenizeError
 */

/**
 * Result from consumeKeyword.
 * @typedef {Object} ConsumeResult
 * @property {number} column - Updated column position
 * @property {number} line - Line number (unchanged for keywords)
 * @property {number} pos - Updated position in the string
 * @property {Token} token - The consumed keyword token
 */

/**
 * Attempts to consume a keyword (true, false, null).
 * @param {string} jsonc - The input string
 * @param {number} startPos - Starting position
 * @param {number} len - Length of input
 * @param {number} line - Current line number
 * @param {number} startColumn - Starting column
 * @param {TokenizeError[]} errors - Errors array to append to
 * @returns {ConsumeResult|null} The result or null if no keyword matches
 */
export function consumeKeyword(jsonc, startPos, len, line, startColumn, errors) {
    // Check for 'true'.
    if (startPos + 4 <= len && jsonc.slice(startPos, startPos + 4) === 'true') {
        return {
            column: startColumn + 4,
            line,
            pos: startPos + 4,
            token: { category: 'keyword', column: startColumn, line, value: 'true' },
        };
    }

    // Check for 'false'.
    if (startPos + 5 <= len && jsonc.slice(startPos, startPos + 5) === 'false') {
        return {
            column: startColumn + 5,
            line,
            pos: startPos + 5,
            token: { category: 'keyword', column: startColumn, line, value: 'false' },
        };
    }

    // Check for 'null'.
    if (startPos + 4 <= len && jsonc.slice(startPos, startPos + 4) === 'null') {
        return {
            column: startColumn + 4,
            line,
            pos: startPos + 4,
            token: { category: 'keyword', column: startColumn, line, value: 'null' },
        };
    }

    // No keyword matched.
    return null;
}

File: wasm/jsonc_lexer/src/consume_keyword.rs
//! Consumes a JSON keyword from JSONC input.

use crate::types::{ConsumeResult, Token};

/// Attempts to consume a keyword (true, false, null).
/// Returns None if no keyword matches.
pub fn consume_keyword(
    bytes: &[u8],
    start_pos: usize,
    len: usize,
    line: u32,
    start_column: u32,
) -> Option<ConsumeResult> {
    // Check for 'true'.
    if start_pos + 4 <= len && &bytes[start_pos..start_pos + 4] == b"true" {
        return Some(ConsumeResult {
            column: start_column + 4,
            line,
            pos: start_pos + 4,
            token: Token {
                category: "keyword".to_string(),
                column: start_column,
                line,
                value: "true".to_string(),
            },
        });
    }

    // Check for 'false'.
    if start_pos + 5 <= len && &bytes[start_pos..start_pos + 5] == b"false" {
        return Some(ConsumeResult {
            column: start_column + 5,
            line,
            pos: start_pos + 5,
            token: Token {
                category: "keyword".to_string(),
                column: start_column,
                line,
                value: "false".to_string(),
            },
        });
    }

    // Check for 'null'.
    if start_pos + 4 <= len && &bytes[start_pos..start_pos + 4] == b"null" {
        return Some(ConsumeResult {
            column: start_column + 4,
            line,
            pos: start_pos + 4,
            token: Token {
                category: "keyword".to_string(),
                column: start_column,
                line,
                value: "null".to_string(),
            },
        });
    }

    // No keyword matched.
    None
}
